### ISSUE TITLE
for convolutional seq 2 seq

### ISSUE THREAD

**Issue by menglin0320** [bentrevett/pytorch-seq2seq#80]
On the inference stage, is there a lot of repeated computation of the same layer on same step. For example, when generating prediction for timestep 1, we do all convolutions for <sos> token, then on second timestep we do the same calculation again.
I just wonder if my understanding is correct.

**Comment by bentrevett** [comment]
Yes, there are repeated calculations during inference. In theory, we could just input the last token generated by the decoder back into the decoder and not the whole generated sequence so far. 

However, as we create the position embeddings on the fly, this would cause them to not align correctly with the generated target sequence and potentially give poor translations (I'll run a few checks with this now).

At the end of the day, these tutorials are designed more for demonstrating how to implement certain models and not really focused on maximizing efficiency. The goal is that once you've read these, you can modify them how you see fit, either to use on your own dataset or to make them more efficient.

**Comment by menglin0320** [comment]
The paper said： Specifically, we pad the input
by k − 1 elements on both the left and right side by zero
vectors, and then remove k elements from the end of the
convolution output.


For your implementation, I feel that remove k elements should be remove k-1 elements. Did they make a typo or I didn't understand it correctly

**Comment by menglin0320** [comment]
So based on my understanding, implement wise, It should be like:
`#tensor conv_decode has shape batch_size, layer, timestep, hiddensize
we only do
conv_decode[:,layer,curtimestep, :] = conv(conv_decode[:,layer-1,  timestep-(k-1):curtimestep+1, : ]) `
we need to apply padding when necessary, but when timestep >k-1 this is the case.

**Comment by bentrevett** [comment]
> The paper said： Specifically, we pad the input
> by k − 1 elements on both the left and right side by zero
> vectors, and then remove k elements from the end of the
> convolution output.
> 
> For your implementation, I feel that "remove k elements" should be "remove k-1 elements". Did they make a typo or I didn't understand it correctly

This padding they're talking about is for the decoding to ensure the model cannot "cheat" and look at future tokens when receiving the entire target sequence in parallel. I'm not sure how this is related to the original post of this issue.

Also, their padding is slightly different from ours as they pad on both sides and remove the convolutional outputs that have only been over padding tokens. We only pad on the left hand side of the target sentence, and thus don't need to remove anything from the right hand side after we convolve.

**Comment by menglin0320** [comment]
This question is not related to the original post. I mean even though the implementation is different, but I guess you guys are trying to achieve the same thing.

What I mean is that there should be k-1 overpadding tokens instead of k overpaddin tokens. Otherwise your implementation is different from the model described in paper.

**Comment by bentrevett** [comment]
OK, so our implementation and the paper do the same thing, but ours does it more efficiently. 

Let's take an example where we have a filter size (k) of 3 and a target sentence with 5 tokens. This is how their padding/decoding happens:

![](https://i.imgur.com/tBwy4y8.jpg)

They pad their sentence on each side by 2 (k-1) padding tokens and then remove the last k outputs. Thus the "last" input is x_4, which should output x_5.

This is how it is done in our implementation:

![](https://i.imgur.com/dgNdSQE.jpg)

We only pad on the left hand side of the sentence and we only put in target tokens up to x_4 into the decoder, see `trg[:,:-1]` in the training loop. This means we don't need to pad the right hand side of the sentence and thus don't need to get rid of some any convolutional outputs. Again, the last "input" is x_4, which should output x_5. 

Thus, our implementation is the same as the paper. Yet theirs is more inefficient as they are doing k calculations and then just getting rid of them.

**Comment by menglin0320** [comment]
get it, thank you very much. 
The thing buffled me is that I thought they pad the token list x1, x2, x3, x4 but in fact they padded x1, x2, x3, x4, x5. 
Thank you for your patience!
